Data Structures/BinaryTree/BT_CountnonLeafNodes.java

```diff
@@ -0,0 +1,119 @@
+package Git;
+
+import java.util.Scanner;
+
+public class BT_CountnonLeafNodes {
+
+	static Scanner scn = new Scanner(System.in);
+
+	public static void main(String[] args) {
+
+		BT_CountnonLeafNodes bts = new BT_CountnonLeafNodes();
+		BinaryTree bt = bts.new BinaryTree();
+		// 10 true 20 true 40 false false true 50 false false true 30 true 60 false
+		// false false **Give this Input to create the tree**
+
+		System.out.println(bt.countNonLeafNodes());
+
+	}
+
+	private class BinaryTree {
+
+		private class Node {
+			int data;
+			Node left;
+			Node right;
+		}
+
+		private Node root;
+		private int size;
+
+		public BinaryTree() {
+			this.root = construct(null, false);
+		}
+
+		private Node construct(Node parent, boolean ilc) {
+
+			if (parent == null) {
+				System.out.println("Enter the data for root node ? ");
+			} else {
+				if (ilc) {
+					System.out.println("Enter the data for left " + parent.data);
+				} else {
+					System.out.println("Enter the data for right " + parent.data);
+				}
+			}
+
+			int val = scn.nextInt();

```

Comment:
No need to ask for input you can just provide a new test input like
`int value=3`

Reply:
This value is for the value of the node being added by the user.
How can you hardcode it with 3.....

Reply:
The node being added by the user might not come from the input of your keyboard. It might be from a network, file, sensor etc. You want you can decouple the input and the processing. Take in the value of 'val' from the scanner supposes user input from the keyboard. This type of algorithm does not scale well. It is better to decouple the input and processing. For your ease, I recommended to hardcode values for demonstration. Even that is not proper practise.